### ISSUE TITLE
iOS - Previous conversation is briefly visible when opening a new chat

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#3729]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open iOS e.cash app and log in
2. Navigate to a conversation
3. Go back to LHN
4. Open a different conversation

## Expected Result:
New chat should be displayed. 

## Actual Result:
Previous conversation is briefly visible before displaying the correct chat. 

## Platform:
Where is this issue occurring?

 Web
 iOS ✔️ 
 Android
 Desktop App
 Mobile Web

**Version Number:** 1.0.73-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 
Issue **not** reproducible in the Android app. 

https://user-images.githubusercontent.com/44479856/123075820-d99a3e00-d418-11eb-9251-d049eadb6d82.MP4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @puneetlath https://expensify.slack.com/archives/C01GTK53T8Q/p1624380109498500
>ISSUE: on iOS staging:
> tap into a chat
> tap back to the home screen
> tap into a different chat
> the last chat you were in will flash on the screen before the new chat shows

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mitchexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by roryabraham** [comment]
Sorry for the delay here, let's get this on Upwork!

**Comment by rdjuric** [comment]
For posterity's sake: We had another issue https://github.com/Expensify/Expensify.cash/issues/2154 open for this before, and a  PR https://github.com/Expensify/Expensify.cash/pull/2221 was made to try and fix it.

I think one of the problems with our current fix is that this code block is never executed the way we intended
https://github.com/Expensify/Expensify.cash/blob/c5707430dff9ecfb631012bb0ba25357c7f367e5/src/pages/home/ReportScreen.js#L37-L44
That's because our component is remounted whenever we change our reportID and the `ComponentDidUpdate` lifecycle [isn't called on the first render](ComponentDidUpdate). This means that in our implementation, every time `ComponentDidUpdate` is called our component will have updated its props already (prevProps === this.props) and our `reportChanged` check will always return false. 


### Proposal
1. Delete the `ComponentDidUpdate` lifecycle
2. Conditionally render our `ReportView` [here](https://github.com/Expensify/Expensify.cash/blob/c5707430dff9ecfb631012bb0ba25357c7f367e5/src/pages/home/ReportScreen.js#L97), something like
`{!this.shouldShowLoader() && <ReportView reportID={this.getReportID()} />}`
3. This gives us the behavior below 

https://user-images.githubusercontent.com/53711423/124070288-f4ba0e80-da13-11eb-8352-d8c2090875e7.mp4

**Comment by roryabraham** [comment]
Thanks for the explanation @rdjuric!

> our component is remounted whenever we change our reportID

It seems like this was not the intention when [this PR](https://github.com/Expensify/Expensify.cash/pull/2221) was written, and I wonder if this is the real source of the issue. Do you know why the `ReportScreen` remounts entirely when the `reportID` changes? I see that the parent component is only passing a [reportID](https://github.com/Expensify/Expensify.cash/blob/55973138355efc9dcb2c5c48ad0efa9c3d839867/src/libs/Navigation/AppNavigator/MainDrawerNavigator.js#L54) as a route param, but why would changing the `route` prop trigger a full remount of the component?

**Comment by rdjuric** [comment]
@roryabraham Good questions!

I don't think the remount is the cause since our `ComponentDidUpdate` would do the same thing as `ComponentDidMount` does in case our `reportChanged` check ever returned true.

i.e remounting is also calling `this.prepareTransition()`  correctly when we change reports, and adding a few logs shows that our [FullScreenLoadingIndicator is also being set](https://github.com/Expensify/Expensify.cash/blob/8e0c6f5c42a417d6515d30d6116bc9659d1c6b74/src/pages/home/ReportScreen.js#L95) to visible as it should.

I think the problem is really that we're trying to render both the `FullScreenLoadingIndicator` and the `ReportView` while loading. So if our `ReportView` takes [300ms](https://github.com/Expensify/Expensify.cash/blob/8e0c6f5c42a417d6515d30d6116bc9659d1c6b74/src/pages/home/ReportScreen.js#L76) or more to be painted, our LoadingIndicator isn't shown at all and we're stuck on the old `ReportView` during the whole transition.

I'll talk a bit about the remounts in the next comment.

**Comment by rdjuric** [comment]
> Do you know why the `ReportScreen` remounts entirely when the `reportID` changes? I see that the parent component is only passing a [reportID](https://github.com/Expensify/Expensify.cash/blob/55973138355efc9dcb2c5c48ad0efa9c3d839867/src/libs/Navigation/AppNavigator/MainDrawerNavigator.js#L54) as a route param, but why would changing the `route` prop trigger a full remount of the component?

It's not the prop change that is triggering the remount (the parent is also being remounted), but the way we navigate in our drawer navigator. This reset here causes the remounts.
https://github.com/Expensify/Expensify.cash/blob/870c3db1675bb294c8961a8db88bc1ffa276dea9/src/libs/Navigation/CustomActions.js#L27-L35

I modified it a bit so that it would stop remounting our routes and this issue kept happening

https://user-images.githubusercontent.com/53711423/124189704-7f901d00-da97-11eb-927a-67da5b1a3ca6.mp4

**Comment by MitchExpensify** [comment]
My bad for the delay! Missed the "external" assignment on this one, here's the [Upwork job](https://www.upwork.com/jobs/~011fda5f8ed18f7863).

**Comment by roryabraham** [comment]
Hmmm okay thanks for digging into this some more @rdjuric. I think that good enough for me to agree that we should just take the path of least resistance and move forward with your [original proposed solution](https://github.com/Expensify/Expensify.cash/issues/3729#issuecomment-871932604). Feel free to submit a PR once you're been hired on Upwork.

@MitchExpensify feel free to hire @rdjuric if/when he applies on Upwork.

**Comment by MitchExpensify** [comment]
Hired!

**Comment by rdjuric** [comment]
Sorry for the delay here @roryabraham! PR is up at #3908

**Comment by roryabraham** [comment]
Merged!

**Comment by mallenexpensify** [comment]
Shouldn't this issue have been closed once the PR was merged or when it hit production?  
@MitchExpensify , can you review, pay in Upwork then close this issue?
I went to do it in Upwork but saw this message from you and got confused
![image](https://user-images.githubusercontent.com/22508304/126360503-d3c729b5-0c2a-4511-aec2-8c33d3d953e8.png)

**Comment by rdjuric** [comment]
> Shouldn't this issue have been closed once the PR was merged or when it hit production?  
> @MitchExpensify , can you review, pay in Upwork then close this issue?
> I went to do it in Upwork but saw this message from you and got confused
> ![image](https://user-images.githubusercontent.com/22508304/126360503-d3c729b5-0c2a-4511-aec2-8c33d3d953e8.png)
> 

I think this is the message I sent to mitch on Upwork 😅

**Comment by MitchExpensify** [comment]
Paid! Apologies for the delay @rdjuric

**Comment by mallenexpensify** [comment]
@rdjuric thanks for comment, makes sense now. 
@arielgreen going back to 'shouldn't this issue have been closed' - Our new process will NOT close the issue so what I should have said was 'shouldn't an automated comment have been added stating that the PR had been merged?  

Also, our new process will be adding `Awaiting payment` and `Weekly` labels to issue once the PR is merged so that contributor managers will know to pay in a week

**Comment by arielgreen** [comment]
@mallenexpensify ah, yeah I think you're correct; _might_ be due to what @MonilBhavsar was saying elsewhere, that since this is an older issue it wasn't in the DB so it didn't get the automated comment.
